### ISSUE TITLE
Caching problem - Storing cache with wrong key and using AddSingleton instead of AddScoped

### ISSUE THREAD

**Issue by lugospod** [AlexTeixeira/Askmethat-Aspnet-JsonLocalizer#42]
Hi, 

According to my interpretation of the code, the library would cache only first language that gets requests. Why? Because JsonStringLocalizer constructor only gets called once. I suppose this is because it is registered (see below) as singleton (AddSingleton) instead of Scoped (AddScoped).  And if the JsonStringLocalizer constructor is called only once, then its base class constructor gets called only once, and that means that only the first langues gets cached. If the you user in the next request changes its option and chooses a different language, that change would not be reflected in the code.

` internal static void AddJsonLocalizationServices(IServiceCollection services)
        {
            services.AddMemoryCache();
            services.AddSingleton<IStringLocalizerFactory, JsonStringLocalizerFactory>();
            services.AddSingleton<IStringLocalizer, JsonStringLocalizer>();
        }`

Another issue is that you are not storing the cache with the same value as you are reading it:
First is concatenation of key + language, and you are storing it only using key and without language.
I added a fix #43 

`          if (!_memCache.TryGetValue($"{CACHE_KEY}_{currentCulture.ThreeLetterISOLanguageName}", out localization))
            {
                ConstructLocalizationObject(_resourcesRelativePath, currentCulture);
                // Set cache options.
                var cacheEntryOptions = new MemoryCacheEntryOptions()
                    // Keep in cache for this time, reset time if accessed.
                    .SetSlidingExpiration(_memCacheDuration);
                // Save data in cache.
                _memCache.Set(CACHE_KEY, localization, cacheEntryOptions);
            }`

**Comment by AlexTeixeira** [comment]
See version 2.1.0-preview-1

**Comment by Compufreak345** [comment]
I'm currently trying to figure out why BaseName seems to be ignored in your current master branch. I think the root cause is in this issue. You are ignoring the baseName at caching, thus loading the initial localization with BaseName "null" and using this cached localization without respecting the BaseName any more.

I think there are 2 approaches to solve this problem:
(1) Use the BaseName in the cached dictionary as key.
(2) Use the BaseName in the cache key.

(1) Would lead to one big dictionary with some redundancies for different base names.
(2) Would lead to many small dictionaries with some redundancies for different base names.

I am currently trying to figure out an option without those redundancies. but I currently can't think of any that wouldn't decrease performance in some way 🤔

**Comment by AlexTeixeira** [comment]
Hi, 

I'm agree with your seconds points, we should lead many small dictionary that will be cached with the baseName as a key.

I think functionality is more important than performance for a first move. Performance can be improved by cleaning code and removing some useless code.

if performance still same as Microsoft packages, it's something :)

**Comment by Compufreak345** [comment]
For now I implemented it like this for option (2) and it seems to work:

```
private string GetCacheKey(CultureInfo ci)
        {
            if (LocalizationOptions.Value.UseBaseName)
            {
                return $"{CACHE_KEY}_{ci.DisplayName}_{BaseName}";
            }
            return $"{CACHE_KEY}_{ci.DisplayName}";
        }
```

Together with my other changes described in #53 the benchmark results on my machine are like this:
``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.437 (1809/October2018Update/Redstone5)
Intel Core i7-6820HQ CPU 2.70GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.202
  [Host]     : .NET Core 2.2.3 (CoreCLR 4.6.27414.05, CoreFX 4.6.27414.05), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.3 (CoreCLR 4.6.27414.05, CoreFX 4.6.27414.05), 64bit RyuJIT


```
|                                          Method |           Mean |         Error |        StdDev |            Min |            Max |     Ratio | RatioSD |   Gen 0 |   Gen 1 |  Gen 2 | Allocated |
|------------------------------------------------ |---------------:|--------------:|--------------:|---------------:|---------------:|----------:|--------:|--------:|--------:|-------:|----------:|
|                                       Localizer |       121.0 ns |      1.585 ns |      1.482 ns |       118.7 ns |       124.2 ns |      1.00 |    0.00 |       - |       - |      - |         - |
|                                   JsonLocalizer |       105.6 ns |      1.236 ns |      1.096 ns |       104.1 ns |       107.1 ns |      0.87 |    0.02 |  0.0113 |       - |      - |      48 B |
|                       JsonLocalizerWithCreation | 1,563,465.3 ns | 32,968.125 ns | 44,011.499 ns | 1,512,751.4 ns | 1,667,846.1 ns | 13,015.77 |  465.56 | 64.4531 | 31.2500 | 3.9063 |  275312 B |
| JsonLocalizerWithCreationAndExternalMemoryCache |     6,182.3 ns |    122.984 ns |    120.787 ns |     5,916.1 ns |     6,381.1 ns |     51.12 |    0.85 |  0.8774 |  0.4349 |      - |    3696 B |
|                JsonLocalizerDefaultCultureValue |       359.4 ns |      7.173 ns |      6.359 ns |       350.2 ns |       370.5 ns |      2.97 |    0.06 |  0.0892 |       - |      - |     376 B |
|                    LocalizerDefaultCultureValue |       387.0 ns |      9.330 ns |      8.727 ns |       378.6 ns |       412.8 ns |      3.20 |    0.09 |  0.0777 |       - |      - |     328 B |

Looks good to me.

**Comment by AlexTeixeira** [comment]
Available since 2.1.0-preview1
